pkg/reconciler/apis/apiexportendpointsliceurls/apiexportendpointslice_reconcile.go

```diff
@@ -0,0 +1,154 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiexportendpointsliceurls
+
+import (
+	"context"
+	"net/url"
+	"path"
+
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog/v2"
+
+	virtualworkspacesoptions "github.com/kcp-dev/kcp/cmd/virtual-workspaces/options"
+	"github.com/kcp-dev/kcp/pkg/logging"
+	apiexportbuilder "github.com/kcp-dev/kcp/pkg/virtual/apiexport/builder"
+	apisv1alpha1 "github.com/kcp-dev/kcp/sdk/apis/apis/v1alpha1"
+	corev1alpha1 "github.com/kcp-dev/kcp/sdk/apis/core/v1alpha1"
+	"github.com/kcp-dev/kcp/sdk/apis/third_party/conditions/util/conditions"
+)
+
+type endpointsReconciler struct {
+	listShards                 func(selector labels.Selector) ([]*corev1alpha1.Shard, error)
+	getAPIExport               func(path logicalcluster.Path, name string) (*apisv1alpha1.APIExport, error)
+	listAPIBindingsByAPIExport func(apiexport *apisv1alpha1.APIExport) ([]*apisv1alpha1.APIBinding, error)
+	shardName                  string
+}
+
+type result struct {
+	url    string
+	remove bool
+}
+
+func (c *controller) reconcile(ctx context.Context, apiExportEndpointSlice *apisv1alpha1.APIExportEndpointSlice) (*result, error) {
+	r := &endpointsReconciler{
+		listShards:                 c.listShards,
+		getAPIExport:               c.getAPIExport,
+		listAPIBindingsByAPIExport: c.listAPIBindingsByAPIExport,
+		shardName:                  c.shardName,
+	}
+
+	return r.reconcile(ctx, apiExportEndpointSlice)
+}
+
+func (r *endpointsReconciler) reconcile(ctx context.Context, apiExportEndpointSlice *apisv1alpha1.APIExportEndpointSlice) (*result, error) {
+	if !conditions.IsTrue(apiExportEndpointSlice, apisv1alpha1.APIExportEndpointSliceReadyForURLs) {
+		return nil, nil
+	}
+
+	s := apiExportEndpointSlice.Status.ShardSelector
+	if s == "" { // should never happen.
+		return nil, nil
+	}
+
+	selector, err := labels.Parse(s)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get APIExport
+	apiExportPath := logicalcluster.NewPath(apiExportEndpointSlice.Spec.APIExport.Path)
+	if apiExportPath.Empty() {
+		apiExportPath = logicalcluster.From(apiExportEndpointSlice).Path()
+	}
+	apiExport, err := r.getAPIExport(apiExportPath, apiExportEndpointSlice.Spec.APIExport.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	shards, err := r.listShards(selector)
+	if err != nil {
+		return nil, err
+	}
+
+	rs, err := r.updateEndpoints(ctx, apiExportEndpointSlice, apiExport, shards)
+	if err != nil {
+		return nil, err
+	}
+
+	return rs, nil
+}
+
+func (r *endpointsReconciler) updateEndpoints(ctx context.Context,
+	apiExportEndpointSlice *apisv1alpha1.APIExportEndpointSlice,
+	apiExport *apisv1alpha1.APIExport,
+	shards []*corev1alpha1.Shard) (*result, error) {
+	logger := klog.FromContext(ctx)
+	var rs result
+	for _, shard := range shards {
+		if shard.Name != r.shardName {
+			continue
+		}
```

Comment:
this is strange. Why watch shards? Why loop here? We know which shard we are.

Reply:
The problem here is that partition might be based on selector:
```
 eu-shards_only: true
```
 We propagate this selector via APIExportEndpointSlice. But now `Shard X`  does not know if I'm on the list of this selector or not. So it needs to list all shards based on selector from status and check "if I'm on the list". 
 
Maybe there is nicer way to do this. need to think